### PR TITLE
Implement P2770R0 "Stashing stashing iterators for proper flattening"

### LINKED ISSUE #3450: P2770R0 Stashing Stashing Iterators For Proper Flattening
WG21-P2770R0 Stashing Stashing Iterators For Proper Flattening

No feature-test macro is mentioned.

### PR DESCRIPTION
**Pull request by cpplearner**
Fixes #3450

### CODE REVIEW

File: stl/inc/regex [hunk 1]
```diff
@@ -2382,6 +2382,7 @@ public:
     using pointer           = const value_type*;
     using reference         = const value_type&;
     using iterator_category = forward_iterator_tag;
+    using iterator_concept  = input_iterator_tag;
```
Comment (frederick-vs-ja) [inline stl/inc/regex:2385]: It seems that we should only add `iterator_concept` if `_HAS_CXX20` (same for `regex_token_iterator`).

File: stl/inc/ranges [hunk 2]
```diff
@@ -3603,14 +3605,28 @@ namespace ranges {
         };
 
         template <bool _Const>
-        class _Iterator
+        class _Iterator_base
+            : public _Category_base<_Maybe_const<_Const, _Vw>, _InnerRng<_Const>, is_reference_v<_InnerRng<_Const>>> {};
+
+        template <bool _Const>
+            requires forward_range<_Vw>
```
Comment (StephanTLavavej) [inline stl/inc/ranges:3612]: Should this say `requires forward_range<_Maybe_const<_Const, _Vw>>` instead? I see that \[range.join.iterator\]'s `OuterIter outer_ = OuterIter();` is being patched to say `present only if Base models forward_range` where `Base` is `maybe-const<Const, V>`.

File: stl/inc/ranges [hunk 3]
```diff
@@ -3603,14 +3605,28 @@ namespace ranges {
...
+        };
+
+        template <bool _Const>
+        class _Iterator : public _Iterator_base<_Const> {
         private:
-            template <bool>
-            friend class _Iterator;
-            template <bool>
-            friend class _Sentinel;
+            friend join_view;
```
Comment (StephanTLavavej) [inline stl/inc/ranges:3627]: This is great (it aligns with what we've been doing in the other views). Suggestion: Should this change also be made to `_Sentinel`, for consistency within this view?

File: stl/inc/ranges [hunk 4]
```diff
@@ -4000,10 +4053,26 @@ namespace ranges {
         };
 
         template <bool _Const>
-        class _Iterator : public _Category_base<_Const> {
+        class _Iterator_base : public _Category_base<_Const> {};
+
+        template <bool _Const>
+            requires forward_range<_Vw>
```
Comment (StephanTLavavej) [inline stl/inc/ranges:4059]: Similarly, should this say `_Maybe_const<_Const, _Vw>` based on how \[range.join.with.iterator\] says `present only if Base models forward_range`?

File: stl/inc/ranges [hunk 5]
```diff
@@ -4016,29 +4085,53 @@ namespace ranges {
...
+                requires forward_range<_Base>
+                : _Mybase{_STD move(_Outer_)}, _Parent{_STD addressof(_Parent_)} {
+                if (this->_Outer_it != _RANGES end(_Parent->_Range)) {
+                    _Inner_it._Emplace_second(_RANGES begin(_Update_inner()));
                     _Satisfy();
                 }
             }
 
-            _NODISCARD constexpr auto&& _Update_inner() {
+            constexpr _Iterator(_Parent_t& _Parent_)
```
Comment (StephanTLavavej) [inline stl/inc/ranges:4099]: This is depicted as `explicit` in the paper.

File: stl/inc/ranges [hunk 6]
```diff
@@ -4062,18 +4155,16 @@ namespace ranges {
...
-                        auto&& _Inner = _Get_inner();
-                        if (_Inner_it._Second != _RANGES end(_Inner)) {
+                        if (_Inner_it._Second != _RANGES end(_Get_inner())) {
                             break;
                         }
 
-                        ++_Outer_it;
-                        if (_Outer_it == _RANGES end(_Parent->_Range)) {
+                        ++_Get_outer();
+                        if (_Get_outer() == _RANGES end(_Parent->_Range)) {
```
Comment (StephanTLavavej) [inline stl/inc/ranges:4167]: Suggestion: Should we bind an `_OuterIter&` here to avoid the repeated calls?

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

**Comment by StephanTLavavej** [comment]
Thanks, this looks great! I have a couple of suggestions, one missing `explicit`, and what I think are a couple of missing `_Maybe_const<_Const, _Vw>` occurrences, but I'm not a ranges expert, so please correct me if I'm wrong. I reviewed everything so with these comments addressed I think this will be good to go.

**Comment by cpplearner** [comment]
It seems that P2770 breaks `iterator(iterator<!Const> i)` when `forward_range<V>` is false and `forward_range<const V>` is true: said constructor initializes `outer_` with `std::move(i.outer_)`, but `i.outer_` is not present in this case.

(P2770R0 correctly points out that `Const` can only be true when `forward_range<const V>` is true, so `outer_` is always present.)

cc @timsong-cpp

**Comment by timsong-cpp** [comment]
> when `forward_range<V>` is false and `forward_range<const V>` is true

That is not a scenario I care about (where's @CaseyCarter's promised paper when you need it?) - I know of no realistic range where adding `const` _strengthens_ the iterator concept (it can make it not-a-range, of course, and there was some musings at one point about _weakening_ to input for views that would otherwise need to cache).

That said, `lazy_split_view` does additionally constrain its const begin/end with `forward_range<V>`, so I'm not opposed to an LWG issue that adds that for consistency.

**Comment by StephanTLavavej** [comment]
There are test failures in `P2441R2_views_join_with` with Clang:

<details>
<summary>Click to expand error output:</summary>

```
Build setup steps:
Build steps:
Build step failed unexpectedly.
Command: "C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-cl.EXE" "D:\a\_work\1\s\tests\std\tests\P2441R2_views_join_with\test.cpp" "-ID:\build\out\inc" "-ID:\a\_work\1\s\llvm-project\libcxx\test\support" "-ID:\a\_work\1\s\tests\std\include" "/nologo" "/Od" "/W4" "/w14061" "/w14242" "/w14265" "/w14582" "/w14583" "/w14587" "/w14588" "/w14749" "/w14841" "/w14842" "/w15038" "/w15214" "/w15215" "/w15216" "/w15217" "/w15262" "/sdl" "/WX" "/Zc:strictStrings" "/D_ENABLE_STL_INTERNAL_CHECK" "/bigobj" "/FIforce_include.hpp" "/w14365" "/D_ENFORCE_FACET_SPECIALIZATIONS=1" "/D_STL_CALL_ABORT_INSTEAD_OF_INVALID_PARAMETER" "/w14640" "/Zc:threadSafeInit-" "/EHsc" "/std:c++latest" "/permissive-" "-fno-ms-compatibility" "-fno-delayed-template-parsing" "-Wno-unqualified-std-cast-call" "/MD" "-m64" "-FeD:\build\tests\std\tests\P2441R2_views_join_with\Output\13\P2441R2_views_join_with.exe" "-link" "-LIBPATH:D:\build\out\lib\amd64" "-LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.35.32213\lib\x64" "/MANIFEST:EMBED"
Exit Code: 1
Standard Error:
--
D:\a\_work\1\s\tests\std\tests\P2441R2_views_join_with\test.cpp(598,42): error: reference to local variable 'expected_ints' declared in enclosing function 'test_lwg3698'
        span<const int> intervals[2] = {{expected_ints + 0, expected_ints + 3}, {expected_ints + 3, expected_ints + 5}};
                                         ^
D:\a\_work\1\s\tests\std\tests\P2441R2_views_join_with\test.cpp(595,9): note: 'expected_ints' declared here
    int expected_ints[] = {1, 2, 3, 5, 7};
        ^
D:\a\_work\1\s\tests\std\tests\P2441R2_views_join_with\test.cpp(598,61): error: reference to local variable 'expected_ints' declared in enclosing function 'test_lwg3698'
        span<const int> intervals[2] = {{expected_ints + 0, expected_ints + 3}, {expected_ints + 3, expected_ints + 5}};
                                                            ^
D:\a\_work\1\s\tests\std\tests\P2441R2_views_join_with\test.cpp(595,9): note: 'expected_ints' declared here
    int expected_ints[] = {1, 2, 3, 5, 7};
        ^
D:\a\_work\1\s\tests\std\tests\P2441R2_views_join_with\test.cpp(598,82): error: reference to local variable 'expected_ints' declared in enclosing function 'test_lwg3698'
        span<const int> intervals[2] = {{expected_ints + 0, expected_ints + 3}, {expected_ints + 3, expected_ints + 5}};
                                                                                 ^
D:\a\_work\1\s\tests\std\tests\P2441R2_views_join_with\test.cpp(595,9): note: 'expected_ints' declared here
    int expected_ints[] = {1, 2, 3, 5, 7};
        ^
D:\a\_work\1\s\tests\std\tests\P2441R2_views_join_with\test.cpp(598,101): error: reference to local variable 'expected_ints' declared in enclosing function 'test_lwg3698'
        span<const int> intervals[2] = {{expected_ints + 0, expected_ints + 3}, {expected_ints + 3, expected_ints + 5}};
                                                                                                    ^
D:\a\_work\1\s\tests\std\tests\P2441R2_views_join_with\test.cpp(595,9): note: 'expected_ints' declared here
    int expected_ints[] = {1, 2, 3, 5, 7};
        ^
4 errors generated.
--
```
</details>

Additionally, after review has begun, it is good to avoid force-pushing PRs - it's not the end of the world but GitHub makes it harder to see what has incrementally changed when this happens. (Force pushes are okay if nobody has started reviewing, or if there is an unusual situation.)

**Comment by hewillk** [comment]
I think you should define an alias to replace `_InnerRng<_Const>`, for example:

> `using _InnerBase =_InnerRng<Const>;`

since `_InnerRng<_Const>` is repeated too many times, it is a bit dazzling.

File: tests/std/tests/P2441R2_views_join_with/test.cpp [hunk 7]
```diff
@@ -558,6 +560,76 @@ void test_valueless_iterator() {
...
+    auto r   = ranges::subrange{stashing_iterator{}, default_sentinel} | views::join_with(views::empty<int>);
+    auto r2  = r;
+    auto it  = r.begin();
+    auto it2 = r2.begin();
+
+    auto itcopy = it;
+    it          = ++it2;
+    assert(*itcopy == 1);
+
+    int expected_ints[]          = {1, 2, 3, 5, 7};
```
Comment (StephanTLavavej) [inline tests/std/tests/P2441R2_views_join_with/test.cpp:595]: These aren't modified, and they're used with `span<const int>`. I strongly believe that these should be `const` or `constexpr`. (Test code doesn't need to adhere to the highest levels of `const` cleanliness, but it's nice to know that our expected values won't change.)

File: tests/std/tests/P2441R2_views_join_with/test.cpp [hunk 8]
```diff
@@ -558,6 +560,76 @@ void test_valueless_iterator() {
...
+        }
+        constexpr const span<const int>* begin() const {
+            return p;
+        }
+        constexpr const span<const int>* end() const {
+            return p + 2;
+        }
+    };
+
+    auto jv  = intricate_range{.p = intervals} | views::join_with(views::empty<int>);
```
Comment (StephanTLavavej) [inline tests/std/tests/P2441R2_views_join_with/test.cpp:615]: Nitpick: I would feel better if this were `jwv` to avoid looking like a copy-paste mistake.

**Comment by StephanTLavavej** [comment]
@hewillk That's pre-existing (24 occurrences) - I would have no objection to introducing such an alias, but I think it should be done in a followup PR because so much code is affected.

**Review by StephanTLavavej — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I've pushed a merge with `main` to resolve a trivial adjacent-add conflict in `tests/libcxx/expected_results.txt`.

I am a greedy kitty who is *speculatively* mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
* I've pushed a merge with `main`, mechanically resolving merge conflicts:
  + In `<ranges>`, kept this PR's changes, but converted braces to parentheses in mem-initializer-lists.
  + In `P0896R4_views_join/test.cpp` and `P2441R2_views_join_with/test.cpp`, resolved trivial adjacent-add conflicts.
* Then, I added a commit for **non-trivial** semantic merge conflict resolutions. #3493 added test coverage for #3014. This defined an `input_iterator` named `init_list_not_constructible_iterator`. Now, this PR is upgrading `join_view` and `join_with_view`'s `const` overload of `begin()` to require `forward_range<const V>`. Therefore, we need to upgrade `init_list_not_constructible_iterator` to be a `forward_iterator`:
  + Mark its `iterator_category` as `forward_iterator_tag`.
  + Make it equality comparable, required for forward iterators.
  + Verify that it is now a `forward_iterator`.
  + Rename usage from `InRange` to `FwdRange`.
* Finally, this found two more improper uses of braces in `<ranges>`, within `join_view::_Iterator_base` and `join_with_view::_Iterator_base`.

**Comment by StephanTLavavej** [comment]
I'm a greedy kitty who wants treats :smirk_cat: and is *speculatively* mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Review by strega-nil-ms — APPROVED** [review]
Looks good! Minor change, but it's pre-existing so no need to reset testing.

File: stl/inc/ranges [hunk 9]
```diff
@@ -3567,37 +3567,36 @@ namespace ranges {
...
 #ifdef __clang__
     template <class _Rng> // TRANSITION, LLVM-47414
-    concept _Can_const_join = input_range<const _Rng> && is_reference_v<range_reference_t<const _Rng>>;
+    concept _Can_const_join = forward_range<const _Rng> && is_reference_v<range_reference_t<const _Rng>>
+                           && input_range<range_reference_t<const _Rng>>;
 #endif // ^^^ workaround ^^^
 
+    template <class _Ty>
+    _NODISCARD constexpr _Ty& _As_lvalue(_Ty&& _Val) noexcept {
+        return static_cast<_Ty&>(_Val);
```
[the remark this answers is not in the document]
Comment (cpplearner) [reply stl/inc/ranges:3576]: IIUC, this `static_cast` was unnecessary before WG21-P2266, but WG21-P2266 makes `_Val` an xvalue in this context, and thus  an explicit cast is now needed.

File: stl/inc/ranges [hunk 10]
```diff
@@ -3567,37 +3567,36 @@ namespace ranges {
 
 #ifdef __clang__
     template <class _Rng> // TRANSITION, LLVM-47414
-    concept _Can_const_join = input_range<const _Rng> && is_reference_v<range_reference_t<const _Rng>>;
+    concept _Can_const_join = forward_range<const _Rng> && is_reference_v<range_reference_t<const _Rng>>
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/ranges:3570]: I know a guy who just created a dozen cleanup PRs - I'll tell him to add this change to one of them. :joy_cat:

**Comment by StephanTLavavej** [comment]
Thanks for implementing this feature, incrementing the repo towards C++23 conformance! :chart_with_upwards_trend: :tada: :heart_eyes_cat:

**Comment by frederick-vs-ja** [comment]
Will this be backported to a future version of VS2019?

It seems that #2727 hasn't been backported to VS2019 16.11.25 yet, so I'm afraid that there's already some ABI-incompatible uses of `join_view` between VS2019 and VS2022.

**Comment by StephanTLavavej** [comment]
At this time, we aren't planning to service VS 2019 for anything that isn't a security bug or a massively impactful bug. It's unfortunate that #2727 was merged days after we [announced the backport](https://devblogs.microsoft.com/cppblog/msvcs-stl-completes-stdc20/) (2022-05-23 vs. 2022-05-10), and while this means that ABI-incompatible uses aren't blocked, I believe that the affected code will be minimal. We completed C++20 and its DRs in VS 2019 but we can't keep changing it forever, and this change just missed the train. Users who are interested in using the most advanced features should really upgrade to VS 2022.

**Pull request merged**